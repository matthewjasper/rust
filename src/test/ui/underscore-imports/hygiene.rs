// Make sure that if an underscore imports have the same hygiene considerations
// as other imports.

#![feature(decl_macro)]

mod x {
    pub use std::ops::Deref as _;
}

macro m($y:ident) {
    mod $y {
        crate::n!();
    }
}

macro n() {
    pub use crate::x::*;
}

macro p() {
    use std::ops::DerefMut as _;
}

m!(y);

fn main() {
    use crate::y::*;
    p!();
    (&()).deref();              //~ ERROR no method named `deref`
    (&mut ()).deref_mut();      //~ ERROR no method named `deref_mut`
}
