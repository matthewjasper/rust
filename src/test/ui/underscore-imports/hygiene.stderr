error[E0599]: no method named `deref` found for type `&()` in the current scope
  --> $DIR/hygiene.rs:29:11
   |
LL |     (&()).deref();
   |           ^^^^^
   |
   = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope, perhaps add a `use` for it:
   |
LL | use std::ops::Deref;
   |

error[E0599]: no method named `deref_mut` found for type `&mut ()` in the current scope
  --> $DIR/hygiene.rs:30:15
   |
LL |     (&mut ()).deref_mut();
   |               ^^^^^^^^^
   |
   = help: items from traits can only be used if the trait is in scope
help: the following trait is implemented but not in scope, perhaps add a `use` for it:
   |
LL | use std::ops::DerefMut;
   |

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0599`.
